include/ensmallen_bits/function.hpp

```diff
@@ -15,7 +16,10 @@
 
 namespace ens {
 
-template<typename FunctionType, typename MatType, typename GradType>
+template<typename FunctionType,
+         typename MatType,
+         typename GradType,
+         typename OptimizerType>
```

Comment:
Do you think it would be easier to have a standalone `Callbacks` class that has static methods of the form

```
  template<typename OptimizerType, typename FunctionType, typename MatType, typename... CType>
  static bool
  Evaluate(const OptimizerType& optimizer,
           const FunctionType& function,
           const MatType& coordinates,
           size_t begin,
           size_t batchSize,
           CType&... callbacks);
```

or something like that?  This would separate the callback infrastructure from the infrastructure that adds missing functions, and might be a little easier to understand.

Reply:
My intention was to reduce the number of lines necessary to use the callbacks, using an extra class would mean one extra function for ``Evaluate`` and `´Gradient``, but I guess since there are some optimisers that don't use the ``Function`` class it might be easier to just go with an extra class? Let me know what you think.

Reply:
Hmm, I think in this case maintainability is a little easier if we separate out the callbacks and function-adding functionality.  And you're right, not all optimizers use `Function`.  In fact, right now, I think it is only needed or useful for differentiable functions or differentiable separable functions.